keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,639 @@
+1. The app `pod` ingests `secret` and runs.
+
+#### Grant Bucket Access (Brownfield)
+
+1. User creates a `Bucket` in their namespace, with reference to a `BucketClass`.
+1. Controller sees the new `Bucket` and applies a `finalizer` for orchestrated deletions.
+1. Controller gets the `BucketClass` referenced by the `Bucket`.
+1. Controller creates a `BucketContent` object with its `BucketClassName` set to the name of its `BucketClass` and a `finalizer`.
+1. Sidecar detects the new `BucketContent` object and gets the associated `BucketClass`.
+1. Sidecar calls the GrantBucketAccess() rpc, passing the `bucketClass.bucketIdentifier` and the `bucketClass.parameters` and is returned a bucket endpoint and credentials.
+1. Sidecar creates a `secret` containing the endpoint and credentials, with a random/unique name and `ownerRef` set to `BucketContent`.
+1. Sidecar updates `BucketContent.secretRef` with its `secret` name and namespace and sets `BucketContent.status.phase` to *“Ready”*.
+1. Controller detects the `BucketContent` update and sees the *“Ready”* phase. 
+1. Controller copies the generated `secret` to the `Bucket` namespace with name `Bucket.secretName`.  The `secret` is created with an ownerRef of the Bucket.
+1. Controller *“binds”* the `Bucket` and `BucketContent` by setting `bucket.status.bucketContentName` and `bucketContent.bucketRef`, and sets both statuses to *“Bound”*.
+1. App `pod` ingests `secret` and runs.
+
+#### Delete Or Revoke Access (Greenfield & Brownfield)
+
+1. The user deletes their `Bucket`, which blocks until the `finalizer` is removed.
+1. The Controller detects the event and sees the `deletionTimestamp` set in `Bucket` and gets the `BucketClass`. 
+1. If the `BucketClass.secretRef` is nil, the object store bucket is not static, and the process continues to step 4.
+1. The Controller deletes the referenced `BucketContent` object, which blocks until the `finalizer` is removed.
+1. The Sidecar detects the `BucketContent` event and sees the `deletionTimestamp`, and gets the referenced `BucketClass`.
+1. If the `BucketClass.bucketIdentifier` is nil, the Sidecar decides the `BucketClass.bucketIdentifier` is a greenfield object store bucket and calls the rpc associated with the `BucketClass.releasePolicy` (*DeleteBucket* or *RevokeBucketAccess*). Otherwise, the `BucketClass.bucketIdentifier` is non-nil, indicating that it is a brownfield object store bucket, and the Sidecar calls the *RevokeBucketAccess()* rpc.
+1. The Sidecar sets `BucketContent.status.phase` to *“Released”.*
+1. The Controller sees `BucketContent` status is “*Released*” and removes `BucketContent`’s `finalizer`.
+1. The `BucketContent` and the dependent `Secret` will be garbage collected.
+1. The Controller removes the `finalizer` from the `Bucket`, allowing it and it’s dependent `secret` to be garbage collected.
+
+### Static Buckets
+
+> Note: No driver, and thus no sidecar, are present to manage provisioning.
```

Comment:
Are static buckets not just a special case of brown field? Or how is it different?

Reply:
Yes. The difference is there is no driver/sidecar so there are more manual steps. The idea with static/driverless is to encourage early COSI adoption where there may not be drivers avail.